### PR TITLE
Add `serde` implementation for `error-stack`

### PR DESCRIPTION
**Pull request by indietyp**
## 🌟 What is the purpose of this PR?

Create a [`serde`](https://docs.rs/serde) implementation for `Report` and `Frame`

## 🔗 Related links

- #793 

## 🔍 What does this change?

- [x] serde implementation for `Report`

## 📜 Does this require a change to the docs?

Yes. The docs need to be updated to make users know that a serde serialization is provided.

## 🐾 Next steps

- [x] Discuss the format used (#793)
- [x] new attachment `serde`
- [x] hook based serialization alternative
- [x] Implement Format
- [ ] ~~PR to implement json hook (follow up)~~
- [ ] Documentation
- [ ] Tests

## 🛡 What tests cover this?

* New unit tests and integration tests are required

## ❓ How to test this?

1. Unit Tests
2. Integration/Snapshot Tests

### CODE REVIEW

File: packages/libs/error-stack/Cargo.toml [hunk 1]
```diff
@@ -35,11 +36,12 @@ rustc_version = "0.2.3"
 
 [features]
 default = ["std", "small"]
-std = ["anyhow?/std", "once_cell?/std"]
+std = ["anyhow?/std", "once_cell?/std", "serde?/std"]
```
Comment (TimDiekmann) [inline packages/libs/error-stack/Cargo.toml:39]: For what the `std` feature of `serde` is required?

Comment (indietyp) [reply packages/libs/error-stack/Cargo.toml:39]: There's no reason right now, just enabled std to keep consistent via weak dependencies but will remove.

Comment (TimDiekmann) [reply packages/libs/error-stack/Cargo.toml:39]: Just noticed that you also specified `once_cell?/std` but `hooks` implies `std`. Could you put the `features = ['std']` on `once_cell` in the dependency section directly (as drive-by)?

**Comment by indietyp** [comment]
This PR implements a hook-based serialization (like #794) and a new attachment kind for `serde` (like `attach_printable`). I am unsure if we should keep both or decide on one of them.

While `serde` is more intuitive, the hook-based serialization allows for greater control but is much more explicit. (hook based would enable us to serialize backtraces easily, which isn't possible otherwise)

Libraries could also enable the `serde` (if wanted) and add their data. At the same time, with the hook-based approach, we would need to explicitly state *every* serde type that we want to serialize. `hook` based gives more control to the end user.

@TimDiekmann, what do you think? Keep both? Only keep `attach_serializable()` or only keep hooks?

**Comment by TimDiekmann** [comment]
The plan is to remove `attach_printable` when `specialization` (`min_specialization` does not allow specialization on traits) has been stabilized. I don't think it's possible to specialize on two different types, if the second type is not a strict subset of the first type, i.e. (abstracting away the actual specialization-structs)

```
fn attach<T>();
fn attach<T: Display>();
fn attach<T: Serialize>();
```

It's not possible, because `T` could be `Display + Serialize` and adding a `T: Display + Serialize` does not help here ([yet?](https://github.com/rust-lang/rust/pull/49624)). Maybe this is possible with negative trait bounds, but as far as I can tell there is no implementation yet.
In these terms, I'd prefer not to add another `attach` function.

**Comment by indietyp** [comment]
I know that `attach_printable` is planned for removal once `specialization` hits. I haven't thought about the ramifications that every type needs to be a strict subset. This, of course, makes this more challenging. 

Therefore I will remove all code relating to `attach_serializable`.

I am pretty new to the concept of `specialization` in Rust, so here's my question:

Would something like this be possible:

```rust
trait SerializeImpl {
   fn serialize(&self) -> Option<&dyn erased_serde::Serialize>;
}

unsafe trait FrameImpl {
    ...
}

impl<C> SerializeImpl for Attachment<C> {
     fn serialize(&self) -> Option<&dyn erased_serde::Serialize> {
       None
     }
}

impl<C> SerializeImpl for Attachment<C> where C: serde::Serialize {
    fn serialize(&self) -> Option<&dyn erased_serde::Serialize> {
      Some(<dyn erased_serde::Serialize>::erase(self.context))
    }
}

unsafe<C> impl FrameImpl for ContextFrame<C> {
   ...
}
```

**Pull request closed without merging**